crates/kubelet/src/store/oci/file.rs

```diff
@@ -1,76 +1,386 @@
 ///
-/// This type is generic over the type of Kubernetes client used
+/// This type is generic over the type of client used
 /// to fetch modules from a remote store. This client is expected
 /// to be a [`Client`]
-pub struct FileStore<C> {
-    root_dir: PathBuf,
-    client: Arc<Mutex<C>>,
-}
+pub type FileStore<C> = CachingStore<FileStorer, C>;
 
-impl<C> FileStore<C> {
+impl<C: Client + Send> FileStore<C> {
     /// Create a new `FileStore`
     pub fn new<T: AsRef<Path>>(client: C, root_dir: T) -> Self {
         Self {
-            root_dir: root_dir.as_ref().into(),
+            storer: Arc::new(RwLock::new(FileStorer {
+                root_dir: root_dir.as_ref().into(),
+            })),
             client: Arc::new(Mutex::new(client)),
         }
     }
+}
+
+pub struct FileStorer {
+    root_dir: PathBuf,
+}
+
+impl FileStorer {
+    /// Create a new `FileStorer`
+    pub fn new<T: AsRef<Path>>(root_dir: T) -> Self {
+        Self {
+            root_dir: root_dir.as_ref().into(),
+        }
+    }
 
     fn pull_path(&self, r: &Reference) -> PathBuf {
         let mut path = self.root_dir.join(r.registry());
         path.push(r.repository());
-        path.push(r.tag());
+        path.push(r.tag().unwrap_or("latest"));
         path
     }
 
     fn pull_file_path(&self, r: &Reference) -> PathBuf {
         self.pull_path(r).join("module.wasm")
     }
 
-    async fn store(&self, image_ref: &Reference, contents: &[u8]) -> anyhow::Result<()> {
-        tokio::fs::create_dir_all(self.pull_path(image_ref)).await?;
-        let path = self.pull_file_path(image_ref);
-        tokio::fs::write(&path, contents).await?;
-        Ok(())
+    fn digest_file_path(&self, r: &Reference) -> PathBuf {
+        self.pull_path(r).join("digest.txt")
     }
 }
 
 #[async_trait]
-impl<C: Client + Send> Store for FileStore<C> {
-    async fn get(&self, image_ref: &Reference) -> anyhow::Result<Vec<u8>> {
+impl Storer for FileStorer {
+    async fn get_local(&self, image_ref: &Reference) -> anyhow::Result<Vec<u8>> {
         let path = self.pull_file_path(image_ref);
         if !path.exists() {
-            debug!(
-                "Image ref '{:?}' doesn't exist on disk. Fetching remotely...",
+            return Err(anyhow::anyhow!(
+                "Image ref {} not available locally",
                 image_ref
-            );
-            let contents = self.client.lock().await.pull(image_ref).await?;
-            self.store(image_ref, &contents).await?;
-            return Ok(contents);
+            ));
         }
 
         debug!("Fetching image ref '{:?}' from disk", image_ref);
         Ok(tokio::fs::read(path).await?)
     }
+    async fn store(&mut self, image_ref: &Reference, image_data: ImageData) -> anyhow::Result<()> {
+        tokio::fs::create_dir_all(self.pull_path(image_ref)).await?;
+        let digest_path = self.digest_file_path(image_ref);
+        if digest_path.exists() {
```

Comment:
I think this can be removed. `tokio::fs::write` completely overwrites the content of a file if it exists

Reply:
No, this is important (and so should maybe be commented).  If the write of the `module.wasm` fails (possibly part way through) - or if the write of the new digest file fails - I do _not_ want to leave the old digest file lying around saying "hey I have this module and its digest is blah."  We have to delete the digest file first, so that the failure mode is "I don't know the digest of this module file, so better to re-fetch than to assume it's any particular digest."

Reply:
Ok, I wondered if that was the case. Let's comment this just to be safe